src/bit-systems/waypoint.ts

```diff
@@ -0,0 +1,209 @@
+  }
+}
+
+function occupyWaypoint(world: HubsWorld, eid: EntityID) {
+  if (!hasComponent(world, Owned, eid)) {
+    throw new Error("Tried to occupy waypoint before owning it.");
+  }
+  NetworkedWaypoint.occupied[eid] = 1;
+  myOccupiedWaypoint = eid;
+}
+
+function nonOccupiableSpawnPoints(world: HubsWorld) {
+  return waypointQuery(world).filter(eid => {
+    const canBeSpawnPoint = Waypoint.flags[eid] & WaypointFlags.canBeSpawnPoint;
+    const canBeOccupied = Waypoint.flags[eid] & WaypointFlags.canBeOccupied;
+    return canBeSpawnPoint && !canBeOccupied && findAncestorWithComponent(world, SceneRoot, eid);
+  });
+}
+
+function occupiableSpawnPoints(world: HubsWorld) {
+  return waypointQuery(world).filter(eid => {
+    const canBeSpawnPoint = Waypoint.flags[eid] & WaypointFlags.canBeSpawnPoint;
+    const canBeOccupied = Waypoint.flags[eid] & WaypointFlags.canBeOccupied;
+    return (
+      canBeSpawnPoint &&
+      canBeOccupied &&
+      !NetworkedWaypoint.occupied[eid] &&
+      findAncestorWithComponent(world, SceneRoot, eid)
+    );
+  });
+}
+
+function* tryOccupyAndSpawn(world: HubsWorld, characterController: CharacterControllerSystem, spawnPoint: EntityID) {
+  moveToWaypoint(world, spawnPoint, characterController, true);
+  takeOwnershipWithTime(world, spawnPoint, 0);
+  occupyWaypoint(world, spawnPoint);
+  // TODO: We could check if we lost ownership, and not wait as long "lostOwnershipWithTimeout"
+  yield sleep(4000);
+  if (entityExists(world, spawnPoint) && hasComponent(world, Owned, spawnPoint)) {
+    takeOwnership(world, spawnPoint);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+function* trySpawnIntoOccupiable(world: HubsWorld, characterController: CharacterControllerSystem) {
+  for (let i = 0; i < 3; i++) {
+    const spawnPoints = occupiableSpawnPoints(world);
+    if (!spawnPoints.length) return false;
+
+    const waypoint = spawnPoints[Math.floor(Math.random() * spawnPoints.length)];
+    if (yield* tryOccupyAndSpawn(world, characterController, waypoint)) return true;
+  }
+
+  return false;
+}
+
+function* moveToSpawnPointJob(world: HubsWorld, characterController: CharacterControllerSystem) {
+  if (yield* trySpawnIntoOccupiable(world, characterController)) return;
+
+  const spawnPoints = nonOccupiableSpawnPoints(world);
+  if (spawnPoints.length) {
+    const waypoint = spawnPoints[Math.floor(Math.random() * spawnPoints.length)];
+    moveToWaypoint(world, waypoint, characterController, true);
+  } else {
+    console.warn("Could not find any available spawn points, spawning at the origin.");
+    characterController.enqueueWaypointTravelTo(new Matrix4().identity(), true, {
+      willDisableMotion: false,
+      willDisableTeleporting: false,
+      snapToNavMesh: true,
+      willMaintainInitialOrientation: false
+    });
+  }
+}
+
+let spawnJob: Coroutine | null = null;
+export function moveToSpawnPoint(world: HubsWorld, characterController: CharacterControllerSystem) {
```

Comment:
From all the calling sites its quite unclear that after this function call the player would not be at the spawn point. The function signature doesn't tell you either since its not async or a generator. 
1. We should probably try and come up with a better name
2. Should we make it return a promise that resolves when the coroutine finishes? We haven't needed it yet so maybe its not useful, but it would make it more clear that its an async process.